options-design/mongo/clientopts/clientopts.go

```diff
@@ -0,0 +1,90 @@
+	DialContext(ctx context.Context, network, address string) (net.Conn, error)
+}
+
+type Option interface {
+	option()
+}
+
+type OptionFunc func(*Client) error
+
+type Client struct{}
+
+type Bundle struct{}
+
+func BundleOptions(...Option) Bundle { return Bundle{} }
+
+func (b *Bundle) Appname(string) *Bundle                                  { return nil }
+func (b *Bundle) AuthMechanism(string) *Bundle                            { return nil }
+func (b *Bundle) AuthMechaismProperties(map[string]string) *Bundle        { return nil }
+func (b *Bundle) AuthSource(s string) *Bundle                             { return nil }
+func (b *Bundle) ConnectTimeout(d time.Duration) *Bundle                  { return nil }
+func (b *Bundle) Dialer(d ContextDialer) *Bundle                          { return nil }
+func (b *Bundle) HeartbeatInterval(d time.Duration) *Bundle               { return nil }
+func (b *Bundle) Hosts(s []string) *Bundle                                { return nil }
+func (b *Bundle) Journal(v bool) *Bundle                                  { return nil }
+func (b *Bundle) LocalThreshold(d time.Duration) *Bundle                  { return nil }
+func (b *Bundle) MaxConnIdleTime(d time.Duration) *Bundle                 { return nil }
+func (b *Bundle) MaxConnsPerHost(u uint16) *Bundle                        { return nil }
+func (b *Bundle) MaxIdleConnsPerHost(u uint16) *Bundle                    { return nil }
+func (b *Bundle) Password(s string) *Bundle                               { return nil }
+func (b *Bundle) ReadConcernLevel(s string) *Bundle                       { return nil }
+func (b *Bundle) ReadPreference(s string) *Bundle                         { return nil }
+func (b *Bundle) ReadPreferenceTagSets(m []map[string]string) *Bundle     { return nil }
+func (b *Bundle) MaxStaleness(d time.Duration) *Bundle                    { return nil }
+func (b *Bundle) ReplicaSet(s string) *Bundle                             { return nil }
+func (b *Bundle) ServerSelectionTimeout(d time.Duration) *Bundle          { return nil }
+func (b *Bundle) Single(v bool) *Bundle                                   { return nil }
+func (b *Bundle) SocketTimeout(d time.Duration) *Bundle                   { return nil }
+func (b *Bundle) SSL(v bool) *Bundle                                      { return nil }
+func (b *Bundle) SSLClientCertificateKeyFile(s string) *Bundle            { return nil }
+func (b *Bundle) SSLClientCertificateKeyPassword(s func() string) *Bundle { return nil }
+func (b *Bundle) SSLInsecure(v bool) *Bundle                              { return nil }
+func (b *Bundle) SSLCaFile(s string) *Bundle                              { return nil }
+func (b *Bundle) WString(s string) *Bundle                                { return nil }
+func (b *Bundle) WNumber(i int) *Bundle                                   { return nil }
+func (b *Bundle) Username(s string) *Bundle                               { return nil }
```

Comment:
There is a credential object defined in the auth specification: https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#mongocredential. It would consolidate a number of these.

Reply:
We could, but I'm not sure how we would want to handle overrides, but I'm not sure how applicable those are to authentication.